opennmt/layers/transformer.py

```diff
@@ -354,15 +480,38 @@ def _compute_kv(x):
 
         cache = (keys, values)
 
+        queries_length = misc.shape_list(queries)[2]
+
+        use_sparse_att = False
+        if self.max_length_full_attention is not None:
+            if memory is not None:
+                raise ValueError("Sparse attention only supports self-attention.")
+            if self.maximum_relative_position is not None:
+                raise ValueError("Sparse attention doesn't support relative positions.")
+            use_sparse_att = queries_length > self.max_length_full_attention
+
+        chunk_length = self.local_attention_radius
         # Dot product attention.
-        dot = tf.matmul(queries, keys, transpose_b=True)
+        if use_sparse_att:
```

Comment:
Maybe you can override `queries`, `keys`, and `values` to make the changes more concise?

Something like,

```python
if use_sparse_att:
    queries, _ = split_chunks(...)
    keys, _ = split_chunks(...)
    values, num_chunks = split_chunks(...)

dot = tf.matmul(queries, keys, transpose_b=True)
...
heads = tf.matmul(drop_attn, values)

if use_sparse_att:
    heads = unchunk(heads, num_chunks)
    heads = heads[:, :, :queries_length, :]
```